source/cppfront.cpp

```diff
@@ -4919,6 +4925,52 @@ class cppfront
 
                 if (decl->is_constructor()) {
                     found_constructor = true;
+
+                    if (n.template_parameters) {
+                        if (auto func = std::get_if<declaration_node::a_function>(&decl->type)) {
+                            const auto& params = (*func)->parameters->parameters;
+                            auto can_generate_ctad = std::equal(
+                                std::cbegin(params)+1, std::cend(params),
+                                std::cbegin(n.template_parameters->parameters), std::cend(n.template_parameters->parameters),
+                                [](const auto& arg, const auto& type) -> bool {
+                                    if (const auto* arg_type = std::get_if<declaration_node::an_object>(&arg->declaration->type)) {
+                                        assert(type->name());
+                                        return (*arg_type)->get_token() && (*arg_type)->get_token()->as_string_view() == type->name()->as_string_view();
+                                    } else {
+                                        return false;
+                                    }
+                                }
+                            );
+                            if (can_generate_ctad) {
+                                ctads.emplace_back();
+                                std::for_each(std::cbegin(params)+1, std::cend(params), [&](const auto& arg) {
+                                    if (const auto* type = std::get_if<declaration_node::an_object>(&arg->declaration->type)) {
+                                        auto& arg_t = *type;
+                                        assert(arg_t->get_token());
+                                        auto type_name = arg_t->get_token()->to_string(true);
+                                        auto arg_type = print_to_string(*arg_t);
+                                        switch (arg->pass) {
+                                            case passing_style::in     : arg_type += " const &"; break;
+                                            case passing_style::out    :
+                                            case passing_style::inout  : arg_type += "&"; break;
```

Comment:
See <https://godbolt.org/z/1dcWs4za1>.
```suggestion
                                            case passing_style::out    : arg_type.insert(0, "cpp2::deferred_init<") += ">*"; break;
                                            case passing_style::inout  : arg_type += "&"; break;
```

Reply:
Oh, thank you. It seems there is only a need to avoid `cpp2::in<T>` as it is the only not deducible type. So, I guess I can make it simpler to reuse the cpp2 way of passing except for passing `in` args.

Reply:
One comment about `data1 = (t = 42);` in the following code:
```cpp
 operator=: (out this, out t: T, inout y: Y, z: Z) = {
        data1 = (t = 42);
        data2 = y;
        data3 = z;
    }
```
I think we shall not support that kind of style of coding - there is very good alternative that works and is more clear about the intentions:
```cpp
    operator=: (out this, out t: T, inout y: Y, z: Z) = {
        t = 42;
        data1 = t;
        data2 = y;
        data3 = z;
    }
```

Reply:
Fixed.

Reply:
Did that work before? Now it errors with:
```output
build/_cppfront/main.cpp:30:18: error: cannot initialize a member subobject of type 'int' with an rvalue of type 'void'
        : data1{ t.construct(42) }
                 ^~~~~~~~~~~~~~~
```